---
title: "Structural Bioinformatics (Pt. 1)"
author: "Vivian Cai"
date: "2/17/2022"
output: pdf_document
---

# 1: Introduction to the RCSB Protein Data Bank (PDB)  

### PDB statistics  

*Download a CSV file from the PDB site (accessible from “Analyze” > “PDB Statistics” > “by Experimental Method and Molecular Type”. Move this CSV file into your RStudio project and use it to answer the following questions:*

#### Q1: What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy.  

As shown below, about **87.2%** of structures in the PDB are solved by X-Ray, and **5.4%** by EM.

```{r}
df <- read.csv("DataExportSummary.csv")
library(dplyr)
df %>%
  summarize(
    percent_xRay = sum(df$"X.ray")/sum(df$"Total"),
    percent_EM = sum(df$"EM")/sum(df$"Total")
    )
```

#### Q2: What proportion of structures in the PDB are protein?  

As shown from the calculations below, about 87.2% of all PDB structures are protein.

```{r}
percent_prot = df[1,8]/sum(df$"Total")
percent_prot
```

#### Q3: Type HIV in the PDB website search box on the home page and determine how many HIV-1 protease structures are in the current PDB?  

There are **1225** structures of HIV-1 protease.

# 2. Visualizing the HIV-1 protease structure  

#### Q4: Water molecules normally have 3 atoms. Why do we see just one atom per water molecule in this structure?  

The PDB files downloaded don't have hydrogen atoms, so we only see the oxygen atom for every water molecule.

#### Q5: There is a conserved water molecule in the binding site. Can you identify this water molecule? What residue number does this water molecule have (see note below)?  

The residue number for the water at the binding site is 308.

#### Optional: Generate and save a figure clearly showing the two distinct chains of HIV-protease along with the ligand. You might also consider showing the catalytic residues ASP 25 in each chain (we recommend Licorice for these side-chains). Upload this figure to Piazza for some extra credit.  

(Posted on piazza)

#### Discussion Topic: Can you think of a way in which indinavir, or even larger ligands and substrates, could enter the binding site?

### Sequence Viewer Extension [OPTIONAL]  

#### Q6: As you have hopefully observed HIV protease is a homodimer (i.e. it is composed of two identical chains). With the aid of the graphic display and the sequence viewer extension can you identify secondary structure elements that are likely to only form in the dimer rather than the monomer?  

The N-terminus of each chain intertwine with the other and forms a beta sheet structure. The same structure is unlikely to form if only one chain is present.

# 3. Introduction to Bio3D in R  

```{r}
library(bio3d)
```

### Reading PDB file data into R  

```{r}
pdb <- read.pdb("1hsg")
```
```{r}
pdb
```


#### Q7: How many amino acid residues are there in this pdb object?  **198**

#### Q8: Name one of the two non-protein residues?  
**MK1**

#### Q9: How many protein chains are in this structure?  **2**

```{r}
# the attributes (+ attr:) of this object are listed 
# on the last couple of lines.
attributes(pdb)
```
```{r}
# To access these individual attributes we use the 
# dollar-attribute name convention that is common 
# with R list objects.
head(pdb$atom)
```

# 4. Comparative structure analysis of Adenylate Kinase  

### Overview  
*Starting from only one Adk PDB identifier (PDB ID: 1AKE) we will search the entire PDB for related structures using BLAST, fetch, align and superpose the identified structures, perform PCA and finally calculate the normal modes of each individual structure in order to probe for potential differences in structural flexibility.*  


### Setup  
```{r}
# Install packages in the R console

# install.packages("bio3d")
# install.packages("ggplot2")
# install.packages("ggrepel")
# install.packages("devtools")
# install.packages("BiocManager")

# BiocManager::install("msa")
# devtools::install_bitbucket("Grantlab/bio3d-view")

```

#### Q10. Which of the packages above is found only on BioConductor and not CRAN?  **msa**

#### Q11. Which of the above packages is not found on BioConductor or CRAN?:  **bio3d-view**

#### Q12. True or False? Functions from the devtools package can be used to install packages from GitHub and BitBucket?  **TURE**

### Search and retrieve ADK structures  

```{r}
library(bio3d)
aa <- get.seq("1ake_A") # chain A only
aa
```


#### Q13. How many amino acids are in this sequence, i.e. how long is this sequence?  **214**

```{r}
# Blast or hmmer search 
b <- blast.pdb(aa)
# Plot a summary of search results
hits <- plot(b)
# List out some 'top hits'
head(hits$pdb.id)

# in case blast doesn't run
# hits <- NULL
# hits$pdb.id <- c('1AKE_A','6S36_A','6RZE_A','3HPR_A','1E4V_A','5EJE_A','1E4Y_A','3X2S_A','6HAP_A','6HAM_A','4K46_A','3GMT_A','4PZL_A')
```
```{r}
# Download releated PDB files
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

### Align and superpose structures
```{r}
# Align releated PDBs
pdbs <- pdbaln(files, fit = TRUE)#, exefile="msa")
```

```{r}
# Vector containing PDB codes for figure axis
ids <- basename.pdb(pdbs$id)

# Draw schematic alignment
plot(pdbs, labels=ids)
```

### Viewing our superposed structures  

```{r}
library(bio3d.view)
library(rgl)

view.pdbs(pdbs)
```

### Annotate collected PDB structures

```{r}
# I couldn't execute this, I kept getting:
# Error in split.default(X, group) : first argument must be a vector

# anno <- pdb.annotate(ids)
# unique(anno$source)
```

### Principal component analysis  

```{r}
# Perform PCA
pc.xray <- pca(pdbs)
plot(pc.xray)
```

```{r}
# Calculate RMSD
rd <- rmsd(pdbs)

# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)

plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```

(Figure 10: Projection of Adenylate kinase X-ray structures. Each dot represents one PDB structure.)

# 5. Optional further visualization  

```{r}
# Visualize first principal component
pc1 <- mktrj(pc.xray, pc=1, file="pc_1.pdb")
```

```{r}
view.xyz(pc1)
```
(Figure 12: Visualization of PC-1 trajectory generated using mktrj().)

```{r}
view.xyz(pc1, col=vec2color( rmsf(pc1) ))
```

```{r}
#Plotting results with ggplot2
library(ggplot2)
library(ggrepel)

df <- data.frame(PC1=pc.xray$z[,1], 
                 PC2=pc.xray$z[,2], 
                 col=as.factor(grps.rd),
                 ids=ids)

p <- ggplot(df) + 
  aes(PC1, PC2, col=col, label=ids) +
  geom_point(size=2) +
  geom_text_repel(max.overlaps = 20) +
  theme(legend.position = "none")
p
```


# 6. Normal mode analysis  

```{r}
# NMA of all structures
modes <- nma(pdbs)
```
```{r}
plot(modes, pdbs, col=grps.rd)
```

#### Q14. What do you note about this plot? Are the black and colored lines similar or different? Where do you think they differ most and why?  

The most flexible regions are the loopy regions (with no secondary structure prediction). The difference between the black and colored lines shows two different conformations of the protein. After nucleotide binding, the loopy regions becomes a lot more dynamic.


